src/mappings/utils/cumulativeCirculatingSupply.ts

```diff
@@ -0,0 +1,329 @@
+                __kind: nativeToken as string,
+            },
+        }
+
+        if (totalIssuanceStorage.isV6) {
+            totalIssuance = await totalIssuanceStorage.getAsV6(param as CurrencyId_V6);
+        } else if (totalIssuanceStorage.isV15) {
+            totalIssuance = await totalIssuanceStorage.getAsV15(param as CurrencyId_V15);
+        } else if (totalIssuanceStorage.isV17) {
+            totalIssuance = await totalIssuanceStorage.getAsV17(param as CurrencyId_V17);
+        } else if (totalIssuanceStorage.isV1020000) {
+            totalIssuance = await totalIssuanceStorage.getAsV1020000(param as CurrencyId_V1020000);
+        } else if (totalIssuanceStorage.isV1021000) {
+            totalIssuance = await totalIssuanceStorage.getAsV1021000(param as CurrencyId_V1021000);
+        } else {
+            ctx.log.warn("UNKOWN STORAGE VERSION: tokens.totalIssuance`");
+            // update height, to avoid more failed lookups for same height
+            totalIssuanceLatest.height = block.height;
+            // return last known value
+            return totalIssuanceLatest.value;
+        }
+    }
+
+    totalIssuanceLatest.height = block.height;
+    totalIssuanceLatest.value = totalIssuance;
+
+    return totalIssuance;
+}
+
+function findEntityBefore(
+    timestampedEntities: CumulativeCirculatingSupply[],
+    beforeTimestamp: Date,
+    customFilter?: (elem: CumulativeCirculatingSupply) => boolean
+): CumulativeCirculatingSupply | undefined {
+    if (timestampedEntities.length < 1) {
+        return undefined;
+    }
+
+    // find the latest entry that has a tillTimestamp of less than the search timestamp
+    const timestampFiltered = timestampedEntities
+        .filter(entity => entity.tillTimestamp.getTime() < beforeTimestamp.getTime());
+
+    // apply custom filter (most likely checking poolId for volume per pool entities)
+    const customFiltered = (customFilter == undefined) ? timestampFiltered : timestampFiltered.filter(customFilter);
+
+    return customFiltered.reduce((prev: CumulativeCirculatingSupply | undefined, current) => {
+        if (prev === undefined) {
+            return current;
+        }
+
+        return prev.tillTimestamp.getTime() >
+            current.tillTimestamp.getTime()
+            ? prev
+            : current;
+    },
+    undefined);
+}
+
+async function recalculateAndSetCirculatingSupply(
+    entity: CumulativeCirculatingSupply
+): Promise<CumulativeCirculatingSupply> {
+    const issuance = entity.totalIssuance;
+    const locked = entity.amountLocked;
+    const reserved = entity.amountReserved;
+    const system = entity.amountSystemAccounts;
+
+    const circulating = issuance - locked - reserved - system;
+    const circulatingHuman = await convertAmountToHuman(entity.currency, circulating);
+
+    entity.amountCirculating = circulating;
+    entity.amountCirculatingHuman = circulatingHuman;
+    return entity;
+}
+
+async function getInitialSupplyValues(
```

Comment:
How do you make sure this is only run for the correct initialization height?

Reply:
By limiting the range (blocks from height X) of most events by providing that as agrument the processor setup.

The range.from values are set here:
https://github.com/interlay/interbtc-squid/pull/124/files#diff-7199f058b35aa392107dbd32106c45331e7f2fa9fc6de120e05e78d9db49f0faR88-R104

And the agruments are used for 6 out of the 7 events we use to track circulating supply here:
https://github.com/interlay/interbtc-squid/pull/124/files#diff-7199f058b35aa392107dbd32106c45331e7f2fa9fc6de120e05e78d9db49f0faR88-R104
Meaning, squid won't invoke those handlers if it is out of range.

The final event, `Tokens.Transfer`, which is also used for other metrics, has an early exit using the range defined above  so it won't invoke the code that tracks circulating supply:
https://github.com/interlay/interbtc-squid/pull/124/files#diff-7484e667befbfd02e6f42d115f9c00e259ac8bcad700edef2daadc2c78df68feR115-R119

For testnet, the range is empty, so all events from any height will be processed. Which effectively means the values will be inconsistent if the running testnet has (had) a runtime before 1024000 and relevant events happen before. I, for one, am okay with that on testnet.

Reply:
Oh, and the second part of the answer (which actually answers your question): 
The initial values are only fetched if there is no existing set of values created and stored during an earlier event.

Meaning only the very first event that tries to recalculate circulating supply will grab the initial set. The ones following it will get the last set of calculated values.